langstream-api-gateway/src/main/java/ai/langstream/apigateway/gateways/LRUTopicProducerCache.java

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ai.langstream.apigateway.gateways;
+
+import ai.langstream.api.runner.code.Record;
+import ai.langstream.api.runner.topics.TopicProducer;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalNotification;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class LRUTopicProducerCache implements TopicProducerCache {
+
+    private static class SharedTopicProducer implements TopicProducer {
+        private final TopicProducer producer;
+        private volatile int referenceCount;
+        private volatile boolean cached = true;
+
+        public SharedTopicProducer(TopicProducer producer) {
+            this.producer = producer;
+        }
+
+        public void acquire() {
+            synchronized (this) {
+                referenceCount++;
+            }
+        }
+
+        public void removedFromCache() {
+            synchronized (this) {
+                cached = false;
+                if (referenceCount == 0) {
+                    producer.close();
+                }
+            }
+        }
+
+        @Override
+        public void start() {
+            producer.start();
+        }
+
+        @Override
+        public void close() {
+            synchronized (this) {
+                referenceCount--;
+                if (referenceCount == 0 && !cached) {
+                    producer.close();
+                }
+            }
+        }
+
+        @Override
+        public CompletableFuture<?> write(Record record) {
+            return producer.write(record);
+        }
+
+        @Override
+        public Object getNativeProducer() {
+            return producer.getNativeProducer();
+        }
+
+        @Override
+        public Object getInfo() {
+            return producer.getInfo();
+        }
+
+        @Override
+        public long getTotalIn() {
+            return 0;
+        }
+    }
+
+    final Cache<Key, SharedTopicProducer> cache;
+
+    public LRUTopicProducerCache(int size) {
+        this.cache =
+                CacheBuilder.newBuilder()
+                        .maximumSize(size)
+                        .expireAfterWrite(10, TimeUnit.MINUTES)
+                        .expireAfterAccess(10, TimeUnit.MINUTES)
+                        .removalListener(
+                                (RemovalNotification<Key, SharedTopicProducer> notification) -> {
+                                    SharedTopicProducer resource = notification.getValue();
+                                    resource.removedFromCache();
+                                })
+                        .build();
+    }
+
+    @Override
+    public synchronized TopicProducer getOrCreate(
```

Comment:
synchronized is not needed here, the cache is already threadsafe, also this would be a point in which many thread will be blocked while on producer is being created